File: golang/01/argus.go
package main

import (
  "fmt"
  "container/heap"
  "bufio"
  "os"
  "strings"
)

type Query struct {
  num, period, next_time int
}

func NewQuery(num int, period int) *Query {
  return &Query{num, period, period}
}

// Creates a new query from string of type "Register <id> <period>"
func NewQueryFromString(s string) *Query {
  var id, period int
  // fmt.Printf("New query from: %s\n", s)
  fmt.Sscanf(s, "Register %d %d", &id, &period)
  return NewQuery(id, period)
}

func (q *Query) NextTime() int {
  q.next_time += q.period
  return q.next_time
}

type PQ []*Query

func (pq PQ) Len() int { return len(pq) }

func (pq PQ) Less(i, j int) bool {
  this, other := pq[i], pq[j]
  if this.next_time == other.next_time {
    return this.num < other.num
  }
  return this.next_time < other.next_time
}

func (pq PQ) Swap(i, j int) {
  pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PQ) Push(x interface{}) {
  *pq = append(*pq, x.(*Query))
}

func (pq *PQ) Pop() interface{} {
  old := *pq
  n := len(old)
  x := old[n-1]
  *pq = old[0 : n-1]
  return x
}

func main() {
  pq := make(PQ, 0)
  reader := bufio.NewReader(os.Stdin)
  // Read lines until '#' is read
  for {
    line, err := reader.ReadString('\n')
    if err != nil {
      panic(err)
    }
    if strings.HasPrefix(line, "#") {
      break
    }
    q := NewQueryFromString(line)
    heap.Push(&pq, q)
  }

  // Read number of queries
  var num_queries int
  line, err := reader.ReadString('\n')
  if err != nil {
    panic(err)
  }
  fmt.Sscanf(line, "%d\n", &num_queries)
  for i := 0; i < num_queries; i++ {
    q := heap.Pop(&pq).(*Query)
    fmt.Printf("%d\n", q.num)
    q.NextTime()
    heap.Push(&pq, q)
  }
}

File: cpp/01/argus.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <queue>

struct Query {
  int num;
  int period;
  int next_time;

  Query(int num, int period, int next_time) : num(num), period(period), next_time(next_time) {}

  /*
   * Creates a new query object from a string of type "Register <id> <period>"
   */
  Query(const std::string& s) {
    std::istringstream iss(s);
    std::string ignore;
    iss >> ignore >> num >> period;
    next_time = period;
  }

  void increase_time() {
    next_time += period;
  }

  // Overload > operator for priority queue
  bool operator>(const Query& other) const {
    if (next_time == other.next_time) {
      return num > other.num;
    }
    return next_time > other.next_time;
  }
};

int main (void) {
  // Store Queries in a priority queue to keep them in order
  // The priority is the period of the query, in ascending order
  std::priority_queue<Query, std::vector<Query>, std::greater<Query>> queries;
  // Read lines from stdin
  std::string line;
  while (std::getline(std::cin, line)) {
    // Exit condition
    if (line[0] == '#') {
      break;
    }
    // Parse line
    Query q(line);
    queries.push(q);
  }

  int n_queries;
  std::cin >> n_queries;

  for (int i = 0; i < n_queries; i++) {
    // Get next query
    Query q = queries.top();
    queries.pop();
    // Print query number
    std::cout << q.num << std::endl;
    // Add query to queue plus period
    q.increase_time();
    queries.push(q);
  }

  return 0;
}
